File: c++/2519 - Postfix Expression Evaluation.cpp
#include <cstdio>
#include <stack>
#include <queue>
#include <cstring>

using namespace std;

int main()
{
	int t;
	scanf("%d%*c", &t);
	while (t--)
	{
		stack<int> p;
		queue<int> c;
		priority_queue<int> pc;
		char cad[101], tam;
		int i, j, x, y, z;
		scanf("%[^\n]%*c", cad);
		tam = strlen(cad);
		for (i = 0; i < tam; i++)
		{
			if (cad[i] != ' ')
			{
				if (cad[i] != '+' && cad[i] != '-' && cad[i] != '*')
				{
					sscanf(cad + i, "%d", &z);
					while (cad[i] >= '0' && cad[i] <= '9')
						i++;
					i--;
					p.push(z);
					c.push(z);
					pc.push((z) * -1);
				}
				else
				{
					switch (cad[i])
					{
					case '+':
						x = p.top();
						p.pop();
						y = p.top();
						p.pop();
						p.push(y + x);

						x = c.front();
						c.pop();
						y = c.front();
						c.pop();
						c.push(y + x);

						x = -pc.top();
						pc.pop();
						y = -pc.top();
						pc.pop();
						pc.push((y + x) * -1);
						break;
					case '-':
						x = p.top();
						p.pop();
						y = p.top();
						p.pop();
						p.push(y - x);

						x = c.front();
						c.pop();
						y = c.front();
						c.pop();
						c.push(y - x);

						x = -pc.top();
						pc.pop();
						y = -pc.top();
						pc.pop();
						pc.push((y - x) * -1);
						break;
					default:
						x = p.top();
						p.pop();
						y = p.top();
						p.pop();
						p.push(y * x);

						x = c.front();
						c.pop();
						y = c.front();
						c.pop();
						c.push(y * x);

						x = -pc.top();
						pc.pop();
						y = -pc.top();
						pc.pop();
						pc.push(y * x * -1);
						break;
					}
				}
			}
		}
		printf("%d %d %d\n", p.top(), c.front(), -pc.top());
	}
	return 0;
}

File: Java/2519 - Postfix Expression Evaluation.java
import java.util.*;

class Main {

  Stack<Integer> stack;
  Queue<Integer> queue;
  PriorityQueue<Integer> pQueue;

  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);
    String line;
    String values[];
    int cases = in.nextInt();

    Main m = new Main();

    in.nextLine();
    while (cases-- != 0) {
      m.stack = new Stack<Integer>();
      m.queue = new LinkedList<Integer>();
      m.pQueue = new PriorityQueue<Integer>();
      line = in.nextLine();
      values = line.split(" ");
      for (String s : values) {
        if (s.equals("-")) {
          m.process('-');
        } else if (s.equals("+")) {
          m.process('+');
        } else if (s.equals("*")) {
          m.process('*');
        } else {
          m.addValue(Integer.parseInt(s));
        }
      }
      System.out.println(
        m.stack.peek() + " " + m.queue.peek() + " " + m.pQueue.peek()
      );
    }
  }

  private void addValue(int val) {
    stack.push(val);
    queue.offer(val);
    pQueue.offer(val);
  }

  private void process(char op) {
    int xs, ys, xq, yq, xp, yp;
    xs = stack.pop();
    ys = stack.pop();

    xq = queue.poll();
    yq = queue.poll();

    xp = pQueue.poll();
    yp = pQueue.poll();
    switch (op) {
      case '-':
        stack.push(ys - xs);
        queue.offer(yq - xq);
        pQueue.offer(yp - xp);
        break;
      case '+':
        stack.push(ys + xs);
        queue.offer(yq + xq);
        pQueue.offer(yp + xp);
        break;
      case '*':
        stack.push(ys * xs);
        queue.offer(yq * xq);
        pQueue.offer(yp * xp);
        break;
    }
  }
}
